sdk/identity/azure-identity/azure/identity/aio/_credentials/vscode_credential.py

```diff
@@ -0,0 +1,58 @@
+# ------------------------------------
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+# ------------------------------------
+from typing import TYPE_CHECKING
+from ..._exceptions import CredentialUnavailableError
+from .._credentials.base import AsyncCredentialBase
+from ..._constants import AZURE_VSCODE_CLIENT_ID
+from .._internal.aad_client import AadClient
+from ..._credentials.vscode_credential import get_credentials
+if TYPE_CHECKING:
+    # pylint:disable=unused-import,ungrouped-imports
+    from typing import Any, Iterable, Optional
+    from azure.core.credentials import AccessToken
+
+class VSCodeCredential(AsyncCredentialBase):
+    """Authenticates by redeeming a refresh token previously saved by VS Code
+
+        """
+    def __init__(self, **kwargs):
+        self._client = kwargs.pop("_client", None) or AadClient("organizations", AZURE_VSCODE_CLIENT_ID, **kwargs)
+
+    async def __aenter__(self):
+        if self._client:
+            await self._client.__aenter__()
+        return self
+
+    async def close(self):
+        """Close the credential's transport session."""
+
+        if self._client:
+            await self._client.__aexit__()
+
+    async def get_token(self, *scopes, **kwargs):
+        # type: (*str, **Any) -> AccessToken
+        """Request an access token for `scopes`.
+
+        .. note:: This method is called by Azure SDK clients. It isn't intended for use in application code.
+
+        When this method is called, the credential will try to get the refresh token saved by VS Code. If a refresh
+        token can be found, it will redeem the refresh token for an access token and return the access token.
+
+        :param str scopes: desired scopes for the access token. This method requires at least one scope.
+        :rtype: :class:`azure.core.credentials.AccessToken`
+        :raises ~azure.identity.CredentialUnavailableError: fail to get refresh token.
+        """
+        if not scopes:
+            raise ValueError("'get_token' requires at least one scope")
+
+        refresh_token = get_credentials()
+        if not refresh_token:
+            raise CredentialUnavailableError(
+                message="No Azure user is logged in to Visual Studio Code."
+            )
+        token = self._client.get_cached_access_token(scopes)
```

Comment:
Shouldn't this go before trying to get a refresh token?

Reply:
If we fail to get a refresh token (which means the user may logged out), why we should still return a valid token if there is one stored?

Reply:
The point is perhaps moot because `get_token` is usually called by an auth policy which caches access tokens itself. There will typically be a period of minutes--during which authorization will succeed--between a user signing out and the next `get_token` call.

Still, you ask a good question. Here's a couple things to think about in answering it:
1. Access tokens aren't connected to refresh tokens. They're valid for their (short) lifetimes even after the refresh token used to acquire them expires or is revoked.
1. In this scenario the user has signed out but left a program running. The program has tried to send a service request. Which is more surprising to the user: the request succeeding, or failing?

There's another interesting case here. The user can change the signed in identity between two calls to `get_token`. That suggests an argument for having `get_token` always redeem the refresh token, if any, for a new access token, although I don't think users should expect changing identities midstream to pass without error. Redeeming the refresh token every time also raises another issue: when we redeem a refresh token, AAD may return us a new one and invalidate the one we sent. We don't contribute this new refresh token back to VS Code. That means our credential can invalidate the Azure Account extension's refresh token but itself continue to succeed and I guess we then recurse on "what happens if the user signs out of the extension?" 😸 

I'm trying to get all the food for thought here down on this page, sorry for all the words. Personally I now lean toward:
1. forget caching in this credential; leave it up to the auth policy
1. if the user signs out or changes the signed in identity, we make no promises: undefined behavior ahead
1. iterate following feedback

Anyone else have thoughts?

Reply:
> In this scenario the user has signed out but left a program running. The program has tried to send a service request. Which is more surprising to the user: the request succeeding, or failing?

I think the choice it failing immediately or failing in a few minutes. I lean toward failing immediately because
1. It is deterministic
2. It is easier to diagnose.